src/simulator/engine/d3-simulator-engine.ts

```diff
@@ -161,9 +165,11 @@ export type D3SimulatorEvents = {
 };
 
 export class D3SimulatorEngine extends Emitter<D3SimulatorEvents> {
-  protected readonly linkForce: ForceLink<ISimulationNode, SimulationLinkDatum<ISimulationNode>>;
-  protected readonly simulation: Simulation<ISimulationNode, undefined>;
-  protected readonly settings: ID3SimulatorEngineSettings;
+  protected linkForce!: ForceLink<ISimulationNode, SimulationLinkDatum<ISimulationNode>>;
+  protected simulation!: Simulation<ISimulationNode, undefined>;
+  // TODO(dlozic): Question: I removed readonly here since I'm using _initialSettings to reassign
+  // the settings in resetSettings(). Is this okay? Should I use Object.assign()?
```

Comment:
Hmm it is a bit dirty, I think you can resolve both with this approach:

- always set `initialSettings`, currently initial settings are set only if there is constructor `settings`. You could have something like this:

```
// constructor
this.initialSettings = Object.assing(copyObject(DEFAULT_SETTINGS), settings ?? {});
```

You wan't to skip having `var: T | undefined` if it doesn't need to be undefined. By checking your code, having `initialSettings` as undefined doesn't make sense because even in `resetSettings` you are again using DEFAULT_SETTINGS instead of initialSettings. So, let's do all of that at once in the constructor. With that, your `initialSettings` will never be `undefined`.

- reset settings can be this:

```
resetSettings() {
  return copyObject(this._initialSettings);
}
```

- in constructor your settings can be then:

```
this.settings = resetSettings();
// or even
this.settings = copyObject(this._initialSettings);
```

What I want to say, whoever you want to reuse a function in the constructor, then make that function in a sense that it returns updated values, so you can do this in the constructor:

```
this.var = myFunc();
```

With that, you don't need to have `!` at the end of the name then.

Additionally, it is way better code, because if someone forgets to set a value in `myFunc()` (like now), because of the `variable!` definition, TS can't catch that.

Reply:
If I have `_initialSettings` always defined (e.g. `DEFAULT_SETTINGS`) then they lose their purpose. I use this variable to store the initially defined settings. So when someone resets the settings, they don't default to `DEFAULT_SETTINGS`, but rather their initially defined settings. _(maybe this is not necessary, but I had a use case for this)_

Why? Because, in a worker context, we need to send the settings via a message - it can't be in the `OrbView` constructor. In `orb-view.ts` we can't provide settings during the creation of `d3-simulator-engine` because we can't send that object to the worker when the worker creates the `D3SimulatorEngine` in a separate thread.

```TypeScript
  const orb = new Orb.OrbView(container);
  orb.setSettings({
    simulation: {
    collision: {
      ...
     },
  });
```

So now, If they are always defined, I need to introduce another variable - `areSettingsInitialized` or something similar to know if the settings have been initialized. Maybe this is cleaner? What do you think?

I remember not being satisfied with this pattern but I implemented it this way because of the use case above.

Reply:
Essentially, we don't have a way for the user to provide their own initial/default settings. The way I implemented it, the settings that he first assigns are treated as initial/default settings.

But now after writing this out, I guess he can simply keep a constant variable `const MY_DEFAULTS` and use `setSettings(MY_DEFAULTS)` instead of us trying to make `resetSettings` work.